### PR TITLE
Added snesing of starting ring & changed package structure

### LINKED ISSUE #44: Ring Detection for Auto
Where to put wobble goal???? Maybe include enum to return which wobble goal is the right one.

### PR DESCRIPTION
**Pull request by chop0**
Resolves #44

### CODE REVIEW

**Comment by chop0** [comment]
Closes #44

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+            ring_angle = 0;
+        } else {
+            double ring_angle_sine;
+            double inches_height = height * inches_per_pixel;
+            double c_constant = 0.75 - inches_height;
+            double[] roots = roots(-0.75, 5, c_constant);
+            double sine_plus = roots[0], sine_minus = roots[1];
+            if (1 > sine_plus && sine_plus > -1) {
+                ring_angle_sine = sine_plus;
+            } else ;
```
Comment (e3l) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java:55]: what in the world is this semicolon

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+                maxVal = contourArea;
+                maxValIdx = contourIdx;
+            }
+        }
+
+        return maxValIdx;
+    }
+
+    static int closestContour(final List<MatOfPoint> contours, Point centre, double area) {
+        double closestDistToCentre = 999999999;
```
Comment (e3l) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java:88]: Double.MAX_VALUE is preferred

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    static final Scalar BLUE = new Scalar(255, 0, 0);
+
+    public enum TARGET_ZONE {
+            TARGET_ZONE_A,
+        TARGET_ZONE_B,
+        TARGET_ZONE_C
+    }
+
+
+    private static double[] roots(double a, double b, double c) {
```
Comment (e3l) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java:25]: See if you can use our existing MathFunctions instead.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                closestIdx = currIdx;
+                closestDistToCentre = dist;
+            }
+            currIdx++;
+        };
+
+        return closestIdx;
+    }
+
+    static double euclideanDist(Point p, Point q) {
```
Comment (e3l) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java:108]: Same idea as above -- let's keep Math in one place if possible.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+import static org.opencv.imgproc.Imgproc.*;
+
+public class RingStackLocator {
+
+    static final Scalar RED = new Scalar(0, 0, 255);
+    static final Scalar GREEN = new Scalar(0, 255, 0);
+    static final Scalar BLUE = new Scalar(255, 0, 0);
+
+    public enum TARGET_ZONE {
```
Comment (e3l) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java:18]: Enums should be named in the same way as classes. The states of the enum should be capital, but not the enum itself.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+
+    private static double[] roots(double a, double b, double c) {
+        double[] roots = new double[2]; //This is now a double, too.
+        double discriminator = Math.sqrt(Math.pow(b, 2) - 4 * a * c);
+        roots[0] = (-b + discriminator) / (2 * a);
+        roots[1] = (-b - discriminator) / (2 * a);
+        return roots;
+    }
+    public static double find_ring_angle(double width, double height) {
```
Comment (e3l) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java:32]: uh could you please leave comments so I don't have to find the intuition behind all this math, much easier for me to read.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/opmodes/VisionTest.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                //webcam.closeCameraDevice();
+            }
+
+            sleep(100);
+        }
+    }
+
+    class VisionTestPipeline extends OpenCvPipeline
+    {
+        public RingStackLocator.TARGET_ZONE numRings = RingStackLocator.TARGET_ZONE.TARGET_ZONE_A;
```
Comment (e3l) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/opmodes/VisionTest.java:60]: Misleading variable name? Also why is this hard coded

**Review by e3l — CHANGES_REQUESTED** [review]

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java (hunk 6 above)
Comment (chop0) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/vision/RingStackLocator.java:32]: will remove find_ring_angle is not useful for now

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/opmodes/VisionTest.java (hunk 7 above)
Comment (chop0) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/opmodes/VisionTest.java:60]: default value so telemetry doesn't throw a nullpointerexception

Comment (e3l) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/opmodes/VisionTest.java:60]: Perhaps add an Unknown state?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/util/auto/MathFunctions.java [hunk 8]
```diff
@@ -203,7 +203,22 @@ public static Point twoLineIntersectionPoint(Point firstLinePoint, double firstL
...
+     * Takes two coordinates (x1, y1) and (x2, y2) and returns the euclidean distance between them, found using the pythagoras theorem:
+     * x^2+y^2=h^2 -> h=sqrt(x^2+y^2)
+     * Where x is the difference between x1 and x2, and y is the difference between y1 and y2.
+     * @param x1
+     * @param y1
+     * @param x2
+     * @param y2
+     * @return The distance between the specified coordinates
+     */
+    public static double euclideanDist(double x1, double y1, double x2, double y2) {
```
Comment (e3l) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/rework/util/auto/MathFunctions.java:219]: uh pretty sure it's Math.hypot. Check the usages in PathFollow

**Review by e3l — CHANGES_REQUESTED** [review]
Kinda want to know what Moments, I'll try to get opencv working tomorrow morning.

**Review by e3l — DISMISSED** [review]
lgtm! remember to rebase + merge. lmk if you need help doing it from command line

**Review by e3l — DISMISSED** [review]
merges lgtm

**Comment by chop0** [comment]
bugger

**Comment by chop0** [comment]
forgot to add an import

**Review by e3l — APPROVED** [review]
lgtm

**Pull request merged**